contracts/api3-server-v1/proxies/Api3CrossRateReaderProxyV1.sol

```diff
@@ -0,0 +1,195 @@
+    CalculationType public immutable override calculationType;
+
+    /// @notice Derived dAPI name of the cross rate as a bytes32 string
+    /// @dev For example if the two proxies are "USD/USDT" and "USDT/EUR", the
+    /// derived dAPI name should be "USD/EUR". This is just for informational
+    /// purposes and is not used in any calculations
+    bytes32 public immutable override crossRateDapiName;
+
+    /// @dev Parameters are validated by Api3CrossRateReaderProxyV1Factory
+    /// @param proxy1_ First IApi3ReaderProxy contract address
+    /// @param proxy2_ Second IApi3ReaderProxy contract address
+    /// @param calculationType_ Calculation type to be used for the cross rate
+    /// @param crossRateDapiName_ dAPI name of the cross rate as a bytes32 string
+    constructor(
+        address proxy1_,
+        address proxy2_,
+        CalculationType calculationType_,
+        bytes32 crossRateDapiName_
+    ) {
+        proxy1 = proxy1_;
+        proxy2 = proxy2_;
+        calculationType = calculationType_;
+        crossRateDapiName = crossRateDapiName_;
+        _disableInitializers();
+    }
+
+    /// @notice Initializes the contract with the initial owner
+    /// @param initialOwner Initial owner
+    function initialize(address initialOwner) external override initializer {
+        __Ownable_init(initialOwner);
+    }
+
+    /// @notice Returns the current value and timestamp of the cross rate between
+    /// two API3 data feeds associated with the two IApi3ReaderProxy contracts
+    /// @dev The value is calculated based on the calculation type and the
+    /// timestamp is the earliest of the two timestamps to ensure that the data
+    /// is not stale
+    /// @return value Cross rate between the two data feed values
+    /// @return timestamp Timestamp of the oldest data feed update
+    function read()
+        public
+        view
+        override
+        returns (int224 value, uint32 timestamp)
+    {
+        (int224 value1, uint32 timestamp1) = IApi3ReaderProxy(proxy1).read();
+        (int224 value2, uint32 timestamp2) = IApi3ReaderProxy(proxy2).read();
+
+        int256 val1 = int256(value1);
+        int256 val2 = int256(value2);
+        int256 crossRate;
+        if (calculationType == CalculationType.Divide1By2) {
+            if (val2 == 0) revert ProxyReturnedZero(proxy2);
+            crossRate = (val1 * 1e18) / val2;
+        } else if (calculationType == CalculationType.Divide2By1) {
+            if (val1 == 0) revert ProxyReturnedZero(proxy1);
+            crossRate = (val2 * 1e18) / val1;
+        } else if (calculationType == CalculationType.Multiply) {
+            crossRate = (val1 * val2) / 1e18;
+        }
+        if (crossRate > type(int224).max || crossRate < type(int224).min) {
+            revert ValueOutOfRange(crossRate);
+        }
+
+        value = int224(crossRate);
+        timestamp = timestamp1 < timestamp2 ? timestamp1 : timestamp2;
```

Comment:
`block.timestamp` imo

Reply:
There is no good aggregation method, so we can go with `block.timestamp` as well. What I have against it is that it's quite different from the timestamp of `Api3ReaderProxyV1` which feels weird. Also, this also doesn't address the [timestamp check problem](https://docs.api3.org/dapps/integration/contract-integration.html#using-timestamp) that complicates OEV. That said, if we go ahead with `block.timestamp` now, we can more easily migrate the standard proxies in the future.

Reply:
https://github.com/api3dao/contracts/pull/472/commits/29cd894ae47e55c70a6fa057c07f5c52e536bc2f

Reply:
I was thinking more about
> That said, if we go ahead with block.timestamp now, we can more easily migrate the standard proxies in the future.

and it feels to me we could do it now. If it's OK to have `block.timestamp` in this contract, it is OK to have it also for the proxies and we can have create `IApi3ReaderProxyV2` that also returns the `block.timestamp`. The common user complaint would be able about the staleness check, but the workaround here is to still expose the off-chain timestamp the users could validate.

WDYT? (This is still a task independent to this and should be addressed separately).

Reply:
Sounds good 👍🏻 I can tackle this in a separate issue and implement an `Api3ReaderProxyV2` that returns `block.timestamp` and maintains backward compatibility with `IApi3ReaderProxy` to not break current dApp integrations.